bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookieInfoReader.java

```diff
@@ -80,121 +78,278 @@ public String toString() {
+        private boolean running = false;
+        private int queuedType = UNQUEUED;
+
+        private boolean shouldStart() {
+            if (running) {
+                return false;
+            } else {
+                running = true;
+                return true;
+            }
+        }
+
+        /**
+         * Mark pending operation FULL and return true if there is no in-progress operation
+         *
+         * @return True if we should execute a scan, False if there is already one running
+         */
+        public boolean tryStartFull() {
+            queuedType = FULL;
+            return shouldStart();
+        }
+
+        /**
+         * Mark pending operation PARTIAL if not full and return true if there is no in-progress operation
+         *
+         * @return True if we should execute a scan, False if there is already one running
+         */
+        public boolean tryStartPartial() {
+            if (queuedType == UNQUEUED) {
+                queuedType = PARTIAL;
+            }
+            return shouldStart();
+        }
+
+        /**
+         * Gets and clears queuedType
+         */
+        public int getAndClearQueuedType() {
+            int ret = queuedType;
+            queuedType = UNQUEUED;
+            return ret;
+        }
+
+        /**
+         * If queuedType != UNQUEUED, returns true, leaves running equal to true
+         * Otherwise, returns false and sets running to false
+         */
+        public boolean completeUnlessQueued() {
+            if (queuedType == UNQUEUED) {
+                running = false;
+                return false;
+            } else {
+                return true;
+            }
+        }
+    }
+    private final InstanceState instanceState = new InstanceState();
+
     BookieInfoReader(BookKeeper bk,
-                          ClientConfiguration conf,
-                          ScheduledExecutorService scheduler) {
+                     ClientConfiguration conf,
+                     ScheduledExecutorService scheduler) {
         this.bk = bk;
         this.conf = conf;
         this.scheduler = scheduler;
     }
-    void start() {
+
+    public void start() {
         scheduler.scheduleAtFixedRate(new Runnable() {
             @Override
             public void run() {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("Running periodic BookieInfo scan");
+                synchronized (BookieInfoReader.this) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Running periodic BookieInfo scan");
+                    }
+                    try {
+                        Collection<BookieSocketAddress> updatedBookies = bk.bookieWatcher.getBookies();
+                        bookieInfoMap.updateBookies(updatedBookies);
+                    } catch (BKException e) {
+                        LOG.info("Got exception {} while querying bookies from watcher, rerunning after {}s",
```

Comment:
any reason the put exception as the first parameter? it is a common practice in slf4j to put exception as the last parameter, and you don't need '{}' for exceptions.